kratos/includes/master_slave_constraint.h

```diff
@@ -0,0 +1,345 @@
+//
+//
+
+#if !defined(CONSTRAINT_H)
+#define CONSTRAINT_H
+// System includes
+
+// project inclu
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+#include "includes/kratos_flags.h"
+#include "geometries/geometry.h"
+#include "containers/flags.h"
+#include "containers/variable.h"
+#include "containers/variable_component.h"
+#include "containers/vector_component_adaptor.h"
+#include "containers/variable_data.h"
+#include "includes/process_info.h"
+namespace Kratos
+{
+/** \brief Constraint * @class MasterSlaveRelation
+    * @ingroup KratosCore
+    * @brief
+	* A class that implements the interface for different master-slave constraints to be applied on a system.
+    * This is the part that is seen by the user from the python level. Objects of this class are
+    * first class citizens of the modelpart.
+    *
+    * This class allows to add a master-slave constraint which is of the form
+    *
+    * SlaveDofVector = T * MasterDofVector + ConstantVector. (Processing of this is currently not implemented.)
+    *
+    * or
+    *
+    * SlaveDof = weight * MasterDof + Constant
+    *
+    * This class's object will provide its slave, master details and relation matrix between them.
+    *
+    * One can add two MasterSlaveConstraint objects with same slave but different masters and weights.
+    * Consider user adds : SlaveDof = weight1 * MasterDof1 + Constant1
+    *              and   : SlaveDof = weight2 * MasterDof2 + Constant2
+    *
+    * These are later consolidated in the builder and solver to make
+    *                    : SlaveDof = weight1 * MasterDof1 + weight2 * MasterDof2 + Constant1+Constant2
+    *       and then converted to :
+    *                    : SlaveEqID = weight1 * MasterEqId1 + weight2 * MasterEqId2 + Constant1+Constant2
+    * This unique equation is used later on to modify the equation system.
+    */
+class MasterSlaveConstraint :  public IndexedObject, public Flags
+{
+
+  public:
+    /// Pointer definition of DataValueContainer
+    KRATOS_CLASS_POINTER_DEFINITION(MasterSlaveConstraint);
+    typedef IndexedObject BaseType;
+    typedef std::size_t IndexType;
+    typedef Dof<double> DofType;
+    typedef std::vector< DofType::Pointer > DofPointerVectorType;
+    typedef Node<3> NodeType;
+    typedef PointerVectorSet<NodeType, IndexedObject> NodesContainerType;
+    typedef std::vector<std::size_t> EquationIdVectorType;
+
+    typedef double ConstantType;
+    typedef Matrix MatrixType;
+    typedef Vector VectorType;
+    typedef Kratos::Variable<double> VariableType;
+    typedef Kratos::VariableComponent<Kratos::VectorComponentAdaptor<Kratos::array_1d<double, 3>>> VariableComponentType;
+    ///@name Life Cycle
+    ///@{
+
+   /*
+    * Empty Constructor
+    */
+    MasterSlaveConstraint(IndexType Id = 0):IndexedObject(Id), Flags()
+    {
+    }
+
+    /// Destructor.
+    virtual ~MasterSlaveConstraint() override
+    {
+
+    }
+
```

Comment:
Since this class is the baseclass I would say it also needs an assignment operator